### PR TITLE
SNOW-703372: Add indexer for StructType to be on par with pyspark beh…

### PR DESCRIPTION
**Pull request by sfc-gh-evandenberg**
…avior.

Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-703372

2. Please describe how your code solves the related issue.

   Add __getitem__ and __setitem__ to StructType to handle by string, index or slice range for StructType.

### CODE REVIEW

File: src/snowflake/snowpark/types.py [hunk 1]
```diff
@@ -284,6 +284,23 @@ def _to_attributes(self) -> List["expression.Attribute"]:
     def __repr__(self) -> str:
         return f"StructType([{', '.join(repr(f) for f in self.fields)}])"
 
+    def __getitem__(self, item: Union[str, int, slice]) -> StructField:
+        """Access fields by name, index or slice."""
+        if isinstance(item, str):
+            for field in self.fields:
+                if field.name == item:
+                    return field
+            raise KeyError("No StructField named {0}".format(item))
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/types.py:293]:
```suggestion
            raise KeyError(f"No StructField named {item}")
```

We prefer using f-strings in new code

**Review by sfc-gh-aalam — CHANGES_REQUESTED** [review]
LGTM. Just one minor suggestion

File: tests/unit/test_types.py [hunk 2]
```diff
@@ -241,6 +241,39 @@ def test_struct_field_name():
...
+    with pytest.raises(KeyError, match="No StructField named d"):
+        st["d"]
+
+    with pytest.raises(IndexError):
+        st[5]
+
+    with pytest.raises(TypeError):
+        st[5.0]
+
+    with pytest.raises(TypeError):
```
Comment (sfc-gh-aalam) [inline tests/unit/test_types.py:273]: could we also add argument `matches` to check error message matches. eg `with pytest.raises(TypeError, matches="StructType object does not support item assignment")`

File: src/snowflake/snowpark/types.py (hunk 1 above)
Comment (sfc-gh-evandenberg) [reply src/snowflake/snowpark/types.py:293]: Changed

File: tests/unit/test_types.py (hunk 2 above)
Comment (sfc-gh-evandenberg) [reply tests/unit/test_types.py:273]: Done

**Comment by sfc-gh-evandenberg** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by sfc-gh-evandenberg** [comment]
recheck

File: tests/unit/test_types.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply tests/unit/test_types.py:273]: could you add the matches for all `pytest.raises`

Comment (sfc-gh-evandenberg) [reply tests/unit/test_types.py]: Done

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/types.py [hunk 3]
```diff
@@ -284,6 +284,23 @@ def _to_attributes(self) -> List["expression.Attribute"]:
...
+            for field in self.fields:
+                if field.name == item:
+                    return field
+            raise KeyError(f"No StructField named {item}")
+        elif isinstance(item, int):
+            return self.fields[item]  # may throw ValueError
+        elif isinstance(item, slice):
+            return StructType(self.fields[item])
+        else:
+            raise TypeError("StructType items should be strings, integers or slices")
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/types.py:299]: nit: add at the end `, but got {type(item)}`

File: tests/unit/test_types.py [hunk 4]
```diff
@@ -241,6 +241,39 @@ def test_struct_field_name():
     assert sf.column_identifier.name == "integer type"
 
 
+def test_struct_get_item():
+    sfa = StructField("a", IntegerType())
+    sfb = StructField("b", StringType())
+    sfc = StructField("c", LongType())
```
Comment (sfc-gh-sfan) [inline tests/unit/test_types.py:247]: Let's name them `field_a`, `field_b`, `field_c`?

File: tests/unit/test_types.py [hunk 5]
```diff
@@ -241,6 +241,39 @@ def test_struct_field_name():
...
+
+    st = StructType([sfa, sfb, sfc])
```
Comment (sfc-gh-sfan) [inline tests/unit/test_types.py:249]: nit: maybe name it `struct_type`? I don't have a strong opinion on this one.

**Review by sfc-gh-sfan — APPROVED** [review]
LGTM. Only small nits.

File: src/snowflake/snowpark/types.py (hunk 3 above)
Comment (sfc-gh-evandenberg) [reply src/snowflake/snowpark/types.py]: Done

File: tests/unit/test_types.py (hunk 4 above)
Comment (sfc-gh-evandenberg) [reply tests/unit/test_types.py]: Done

File: tests/unit/test_types.py (hunk 5 above)
Comment (sfc-gh-evandenberg) [reply tests/unit/test_types.py]: Done

**Pull request merged**